File: mtgSearch/src/main/java/com/dbottillo/mtgsearchfree/view/fragments/MainFragment.java
package com.dbottillo.mtgsearchfree.view.fragments;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.support.v4.app.ActivityOptionsCompat;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.Animation;
import android.view.animation.Transformation;
import android.widget.AdapterView;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.ListView;
import android.widget.RelativeLayout;
import android.widget.TextView;
import android.widget.Toast;

import com.dbottillo.mtgsearchfree.MTGApp;
import com.dbottillo.mtgsearchfree.R;
import com.dbottillo.mtgsearchfree.model.CardsBucket;
import com.dbottillo.mtgsearchfree.model.DeckBucket;
import com.dbottillo.mtgsearchfree.model.MTGCard;
import com.dbottillo.mtgsearchfree.model.MTGSet;
import com.dbottillo.mtgsearchfree.model.storage.GeneralPreferences;
import com.dbottillo.mtgsearchfree.presenter.CardsPresenter;
import com.dbottillo.mtgsearchfree.presenter.SetsPresenter;
import com.dbottillo.mtgsearchfree.util.AnimationUtil;
import com.dbottillo.mtgsearchfree.util.LOG;
import com.dbottillo.mtgsearchfree.util.MaterialWrapper;
import com.dbottillo.mtgsearchfree.util.TrackingManager;
import com.dbottillo.mtgsearchfree.view.CardsView;
import com.dbottillo.mtgsearchfree.view.SetsView;
import com.dbottillo.mtgsearchfree.view.activities.CardsActivity;
import com.dbottillo.mtgsearchfree.view.activities.MainActivity;
import com.dbottillo.mtgsearchfree.view.adapters.GameSetAdapter;
import com.dbottillo.mtgsearchfree.view.adapters.OnCardListener;
import com.dbottillo.mtgsearchfree.view.helpers.CardsHelper;
import com.dbottillo.mtgsearchfree.view.helpers.DialogHelper;
import com.dbottillo.mtgsearchfree.view.views.MTGCardListView;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import butterknife.Bind;
import butterknife.ButterKnife;
import butterknife.OnClick;

public class MainFragment extends BasicFragment implements
        MainActivity.MainActivityListener, OnCardListener, CardsView, SetsView, SortDialogFragment.SortDialogListener {

    @Inject
    CardsPresenter cardsPresenter;
    @Inject
    SetsPresenter setsPresenter;

    private MTGSet gameSet;
    private ArrayList<MTGSet> sets = new ArrayList<>();
    private CardsBucket cardBucket;
    private GameSetAdapter setAdapter;
    private int currentSetPosition = -1;
    MainActivity mainActivity;

    @Bind(R.id.set_arrow)
    ImageView setArrow;
    @Bind(R.id.set_list_bg)
    View setListBg;
    @Bind(R.id.set_list)
    ListView setList;
    @Bind(R.id.set_chooser_name)
    TextView chooserName;
    @Bind(R.id.cards_list_view)
    MTGCardListView mtgCardListView;
    @Bind(R.id.cards_view_type)
    ImageButton viewType;
    @Bind(R.id.main_tooltip)
    View tooltip;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.fragment_main, container, false);
        ButterKnife.bind(this, rootView);
        return rootView;
    }

    @Override
    public void onViewCreated(final View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        setActionBarTitle(getString(R.string.app_long_name));

        getMTGApp().getUiGraph().inject(this);
        cardsPresenter.init(this);
        setsPresenter.init(this);

        setAdapter = new GameSetAdapter(getActivity().getApplicationContext(), sets);
        setAdapter.setCurrent(currentSetPosition);
        setList.setAdapter(setAdapter);
        setList.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                if (currentSetPosition != position) {
                    currentSetPosition = position;
                    showHideSetList(true);
                } else {
                    showHideSetList(false);
                }
            }
        });

        view.findViewById(R.id.set_chooser).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                showHideSetList(false);
            }
        });

        GeneralPreferences generalPreferences = GeneralPreferences.with(getContext());
        if (generalPreferences.isTooltipMainToShow()) {
            tooltip.setVisibility(View.VISIBLE);
            MaterialWrapper.setElevation(tooltip, getResources().getDimensionPixelSize(R.dimen.toolbar_elevation));
        } else {
            tooltip.setVisibility(View.GONE);
        }

        setsPresenter.loadSets();
    }

    @Override
    public void onResume() {
        super.onResume();
        cardsPresenter.loadCardTypePreference();
    }

    public void onAttach(Context context) {
        super.onAttach(context);
        mainActivity = (MainActivity) context;
        mainActivity.setMainActivityListener(this);
    }

    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putInt("currentSetPosition", currentSetPosition);
        outState.putParcelableArrayList("SET", sets);
    }

    @OnClick(R.id.cards_sort)
    public void onSortClicked(View view) {
        LOG.d();
        SortDialogFragment sortDialogFragment = new SortDialogFragment();
        sortDialogFragment.show(mainActivity.getSupportFragmentManager(), sortDialogFragment.getTag());
        sortDialogFragment.setListener(this);
    }

    @Override
    public void cardTypePreferenceChanged(boolean grid) {
        LOG.d();
        if (grid) {
            mtgCardListView.setGridOn();
            viewType.setImageResource(R.drawable.cards_list_type);
        } else {
            mtgCardListView.setListOn();
            viewType.setImageResource(R.drawable.cards_grid_type);
        }
    }

    @OnClick(R.id.cards_view_type)
    public void onViewTypeChanged(View view) {
        LOG.d();
        cardsPresenter.toggleCardTypeViewPreference();
    }

    @OnClick(R.id.set_list_bg)
    public void onSetListBgClicked(View view) {
        LOG.d();
        if (setList.getHeight() > 0) {
            showHideSetList(false);
        }
    }

    @OnClick(R.id.main_tooltip_close)
    public void onCloseTooltip(View view) {
        LOG.d();
        GeneralPreferences.with(view.getContext()).setTooltipMainHide();
        AnimationUtil.animateHeight(tooltip, 0);
    }

    private void showHideSetList(final boolean loadSet) {
        LOG.d();
        final int startHeight = setList.getHeight();
        final int targetHeight = ((startHeight == 0)) ? mtgCardListView.getHeight() : 0;
        final float startRotation = setArrow.getRotation();
        final Animation animation = new Animation() {
            public void applyTransformation(float interpolatedTime, Transformation t) {
                super.applyTransformation(interpolatedTime, t);
                if (targetHeight > startHeight) {
                    int newHeight = (int) (startHeight + (interpolatedTime * targetHeight));
                    setHeightView(setList, newHeight);
                    setHeightView(setListBg, newHeight);
                    setArrow.setRotation(startRotation + (180 * interpolatedTime));
                } else {
                    int newHeight = (int) (startHeight - startHeight * interpolatedTime);
                    setHeightView(setList, newHeight);
                    setHeightView(setListBg, newHeight);
                    setArrow.setRotation(startRotation - (180 * interpolatedTime));
                }
            }
        };
        animation.setDuration(200);
        animation.setAnimationListener(new Animation.AnimationListener() {
            public void onAnimationStart(Animation animation) {

            }

            public void onAnimationEnd(Animation animation) {
                if (loadSet) {
                    TrackingManager.trackSet(gameSet, sets.get(currentSetPosition));
                    SharedPreferences.Editor editor = sharedPreferences.edit();
                    editor.putInt("setPosition", currentSetPosition);
                    editor.apply();
                    setAdapter.setCurrent(currentSetPosition);
                    setAdapter.notifyDataSetChanged();
                    loadSet();
                }
            }

            public void onAnimationRepeat(Animation animation) {

            }
        });
        getView().startAnimation(animation);
    }

    private void setHeightView(View view, int value) {
        RelativeLayout.LayoutParams params = (RelativeLayout.LayoutParams) view.getLayoutParams();
        params.height = value;
        view.setLayoutParams(params);
    }

    public void favIdLoaded(int[] favourites) {
        // favourites are not needed in this fragment
        throw new UnsupportedOperationException();
    }

    public void setsLoaded(List<MTGSet> sets) {
        LOG.d();
        currentSetPosition = sharedPreferences.getInt("setPosition", 0);
        setAdapter.setCurrent(currentSetPosition);
        this.sets.clear();
        for (MTGSet set : sets) {
            this.sets.add(set);
        }
        setAdapter.notifyDataSetChanged();
        loadSet();
    }

    public void showError(String message) {
        Toast.makeText(getActivity(), message, Toast.LENGTH_SHORT).show();
    }

    private void loadSet() {
        LOG.d();
        gameSet = sets.get(currentSetPosition);
        chooserName.setText(gameSet.getName());
        cardsPresenter.loadCards(gameSet);
    }

    public void cardLoaded(CardsBucket bucket) {
        LOG.d();
        cardBucket = bucket;
        updateContent();
    }

    @Override
    public void deckLoaded(DeckBucket bucket) {
        throw new UnsupportedOperationException();
    }

    public void updateContent() {
        LOG.d();
        CardsBucket bucket = CardsHelper.filterCards(mainActivity.getCurrentFilter(), cardBucket);
        CardsHelper.sortCards(sharedPreferences, bucket);
        mtgCardListView.loadCards(bucket, this);
    }

    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.main, menu);
        super.onCreateOptionsMenu(menu, inflater);
    }

    public void onSortSelected() {
        LOG.d();
        loadSet();
    }

    @SuppressLint("NewApi")
    @Override
    public void onCardSelected(MTGCard card, int position, View view) {
        LOG.d();
        if (mainActivity.isFilterOpen()) {
            mainActivity.closePanel();
            return;
        }
        TrackingManager.trackCard(gameSet, position);
        Intent intent;
        if (view != null && MTGApp.isActivityTransitionAvailable()) {
            intent = CardsActivity.newInstance(getContext(), gameSet, position, card);
            view.setTransitionName(getString(R.string.transition_card));
            ActivityOptionsCompat activityOptionsCompat = ActivityOptionsCompat.makeSceneTransitionAnimation(getActivity(), view, getString(R.string.transition_card));
            startActivity(intent, activityOptionsCompat.toBundle());
        } else {
            intent = CardsActivity.newInstance(getContext(), gameSet, position, null);
            startActivity(intent);
        }
    }

    @Override
    public void onOptionSelected(MenuItem menuItem, MTGCard card, int position) {
        if (menuItem.getItemId() == R.id.action_add_to_deck) {
            DialogHelper.open(dbActivity, "add_to_deck", AddToDeckFragment.newInstance(card));
        } else {
            cardsPresenter.saveAsFavourite(card, true);
        }
    }

    public String getPageTrack() {
        return "/set";
    }

}

File: mtgSearch/src/main/kotlin/com/dbottillo/mtgsearchfree/view/fragments/MainFragment.kt
package com.dbottillo.mtgsearchfree.view.fragments

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.*
import android.view.animation.Animation
import android.view.animation.Transformation
import android.widget.*
import butterknife.ButterKnife
import butterknife.OnClick
import butterknife.bindView
import com.dbottillo.mtgsearchfree.R
import com.dbottillo.mtgsearchfree.adapters.CardListAdapter
import com.dbottillo.mtgsearchfree.adapters.GameSetAdapter
import com.dbottillo.mtgsearchfree.adapters.OnCardListener
import com.dbottillo.mtgsearchfree.base.MTGApp
import com.dbottillo.mtgsearchfree.cards.CardsHelper
import com.dbottillo.mtgsearchfree.database.CardDataSource
import com.dbottillo.mtgsearchfree.dialog.AddToDeckFragment
import com.dbottillo.mtgsearchfree.helper.DialogHelper
import com.dbottillo.mtgsearchfree.resources.MTGCard
import com.dbottillo.mtgsearchfree.resources.MTGSet
import com.dbottillo.mtgsearchfree.tracking.TrackingManager
import com.dbottillo.mtgsearchfree.util.DialogUtil
import com.dbottillo.mtgsearchfree.view.activities.CardsActivity
import com.dbottillo.mtgsearchfree.view.activities.MainActivity
import fr.castorflex.android.smoothprogressbar.SmoothProgressBar
import java.util.*
import javax.inject.Inject

class MainFragment : BasicFragment(), DialogUtil.SortDialogListener,
        MainActivity.MainActivityListener, OnCardListener, CardsView, SetsView {

    @Inject lateinit var cardsPresenter: CardsPresenter
    @Inject lateinit var setsPresenter: SetsPresenter

    private var gameSet: MTGSet? = null
    private var sets: ArrayList<MTGSet> = ArrayList()
    private var cardBucket: CardsBucket? = null
    private var cards: ArrayList<MTGCard> = ArrayList()
    private var adapter: CardListAdapter? = null
    private var setAdapter: GameSetAdapter? = null
    private var currentSetPosition = -1
    var mainActivity: MainActivity? = null

    val listView: ListView by bindView(R.id.card_list)
    val emptyView: TextView by bindView(R.id.empty_view)
    val setArrow: ImageView by bindView(R.id.set_arrow)
    val setListBg: View by bindView(R.id.set_list_bg)
    val setList: ListView by bindView(R.id.set_list)
    val container: View by bindView(R.id.container)
    val progressBar: SmoothProgressBar by bindView(R.id.progress)
    val chooserName: TextView by bindView(R.id.set_chooser_name)

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val rootView = inflater!!.inflate(R.layout.fragment_main, container, false)
        ButterKnife.bind(this, rootView)
        return rootView
    }

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        setActionBarTitle(getString(R.string.app_long_name))

        adapter = CardListAdapter(activity, cards, false, R.menu.card_option, this)
        listView.adapter = adapter

        MTGApp.dataGraph.inject(this)
        cardsPresenter.init(this)
        setsPresenter.init(this)
        setsPresenter.loadSets();

        setAdapter = GameSetAdapter(activity.applicationContext, sets)
        setAdapter!!.setCurrent(currentSetPosition)
        setList.adapter = setAdapter
        setList.onItemClickListener = AdapterView.OnItemClickListener {
            parent, view, position, id ->
            if (currentSetPosition != position) {
                currentSetPosition = position
                showHideSetList(true)
            } else {
                showHideSetList(false)
            }
        }

        view?.findViewById(R.id.set_chooser)?.setOnClickListener({ showHideSetList(false) })
    }

    override fun onAttach(context: Context?) {
        super.onAttach(context)
        mainActivity = context as MainActivity
        mainActivity?.setMainActivityListener(this)
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putInt("currentSetPosition", currentSetPosition)
        outState.putParcelableArrayList("SET", sets)
    }

    @OnClick(R.id.cards_sort)
    fun onSortClicked(view: View) {
        DialogUtil.chooseSortDialog(context, sharedPreferences, this)
    }

    @OnClick(R.id.set_list_bg)
    fun onSetListBgClicked(view: View) {
        if (setList.height > 0) {
            showHideSetList(false)
        }
    }

    private fun showHideSetList(loadSet: Boolean) {
        val startHeight = setList.height
        val targetHeight = if ((startHeight == 0)) container.height else 0
        val startRotation = setArrow.rotation
        val animation = object : Animation() {
            override fun applyTransformation(interpolatedTime: Float, t: Transformation) {
                super.applyTransformation(interpolatedTime, t)
                if (targetHeight > startHeight) {
                    val newHeight = (startHeight + (interpolatedTime * targetHeight)).toInt()
                    setHeightView(setList, newHeight)
                    setHeightView(setListBg, newHeight)
                    setArrow.rotation = startRotation + (180 * interpolatedTime)
                } else {
                    val newHeight = (startHeight - startHeight * interpolatedTime).toInt()
                    setHeightView(setList, newHeight)
                    setHeightView(setListBg, newHeight)
                    setArrow.rotation = startRotation - (180 * interpolatedTime)
                }
            }
        }
        animation.duration = 200
        animation.setAnimationListener(object : Animation.AnimationListener {
            override fun onAnimationStart(animation: Animation) {

            }

            override fun onAnimationEnd(animation: Animation) {
                if (loadSet) {
                    TrackingManager.trackSet(gameSet, sets[currentSetPosition])
                    val editor = sharedPreferences.edit()
                    editor.putInt("setPosition", currentSetPosition)
                    editor.apply()
                    setAdapter?.setCurrent(currentSetPosition)
                    setAdapter?.notifyDataSetChanged()
                    loadSet()
                }
            }

            override fun onAnimationRepeat(animation: Animation) {

            }
        })
        view!!.startAnimation(animation)
    }

    private fun setHeightView(view: View, value: Int) {
        val params = view.layoutParams as RelativeLayout.LayoutParams
        params.height = value
        view.layoutParams = params
    }

    override fun favIdLoaded(favourites: IntArray) {
        // favourites are not needed in this fragment
        throw UnsupportedOperationException()
    }

    override fun setsLoaded(sets: List<MTGSet>) {
        currentSetPosition = sharedPreferences.getInt("setPosition", 0)
        setAdapter?.setCurrent(currentSetPosition)
        this.sets.clear()
        for (set in sets) {
            this.sets.add(set)
        }
        setAdapter?.notifyDataSetChanged()
        loadSet()
    }

    override fun showError(message: String) {
        Toast.makeText(activity, message, Toast.LENGTH_SHORT).show()
    }

    private fun loadSet() {
        gameSet = sets[currentSetPosition]
        chooserName.text = gameSet?.name
        cardsPresenter.loadCards(gameSet!!)
    }

    override fun luckyCardsLoaded(cards: ArrayList<MTGCard>) {
        throw UnsupportedOperationException()
    }

    override fun cardLoaded(bucket: CardsBucket) {
        cardBucket = bucket
        updateContent()
    }

    override fun updateContent() {
        cards.clear()
        CardsHelper.filterCards(mainActivity?.currentFilter!!, cardBucket?.cards!!, cards)
        val wubrgSort = sharedPreferences.getBoolean(BasicFragment.PREF_SORT_WUBRG, true)
        CardsHelper.sortCards(wubrgSort, cards)

        adapter?.notifyDataSetChanged()
        emptyView.visibility = if (adapter?.count == 0) View.VISIBLE else View.GONE
        listView.smoothScrollToPosition(0)

        if (cards.size == CardDataSource.LIMIT) {
            val footer = LayoutInflater.from(activity).inflate(R.layout.search_bottom, listView, false)
            val moreResult = footer.findViewById(R.id.more_result) as TextView
            moreResult.text = resources.getQuantityString(R.plurals.search_limit, CardDataSource.LIMIT, CardDataSource.LIMIT)
            listView.addFooterView(footer)
        }
        progressBar.visibility = View.GONE

        emptyView.visibility = if (adapter?.count == 0) View.VISIBLE else View.GONE
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.main, menu)
        super.onCreateOptionsMenu(menu, inflater)
    }

    override fun onSortSelected() {
        loadSet()
    }

    override fun onCardSelected(card: MTGCard?, position: Int) {
        TrackingManager.trackCard(gameSet, position)
        val cardsView = Intent(activity, CardsActivity::class.java)
        cardsView.putExtra(CardsActivity.POSITION, position)
        cardsView.putExtra(CardsActivity.KEY_SET, gameSet)
        startActivity(cardsView)
    }

    override fun onOptionSelected(menuItem: MenuItem, card: MTGCard, position: Int) {
        if (menuItem.itemId == R.id.action_add_to_deck) {
            DialogHelper.open(dbActivity!!, "add_to_deck", AddToDeckFragment.newInstance(card))
        } else {
            cardsPresenter.saveAsFavourite(card)
        }
    }

    override fun getPageTrack(): String? {
        return "/set";
    }

}
